File: backend/app/schemas/literature.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.citation import CitationOut


class LiteratureSearchIn(BaseModel):
    query: str = Field(default="", max_length=500)
    rows: int = Field(default=20, ge=1, le=50)
    refined_queries: list[str] | None = None
    must_include: list[str] | None = None
    must_exclude: list[str] | None = None
    skip_refine: bool = False


class LiteratureRefineQueryIn(BaseModel):
    scope: Literal["book", "chapter"] = "book"
    chapter_index: int | None = None
    raw_query: str = Field(default="", max_length=500)


class LiteratureRefineQueryOut(BaseModel):
    refined_queries: list[str] = []
    must_include: list[str] = []
    must_exclude: list[str] = []


class LiteraturePaperOut(BaseModel):
    title: str
    year: int | None = None
    authors: list[str] = []
    journal: str = ""
    doi: str = ""
    citations: int = 0
    type: str | None = None
    source: str | None = None
    source_label: str | None = None
    url: str = ""
    semantic_scholar_id: str | None = None
    external_id: str | None = None
    abstract_preview: str | None = None


class LiteratureSearchOut(BaseModel):
    papers: list[LiteraturePaperOut] = []
    github: list[LiteraturePaperOut] = []
    wiki: list[LiteraturePaperOut] = []
    official_docs: list[LiteraturePaperOut] = []
    refined_queries: list[str] = []
    warnings: list[str] = []
    profile: str = ""
    source_hint: str = ""
    # 兼容旧前端
    items: list[LiteraturePaperOut] = []


class LiteratureQuoteBlockOut(BaseModel):
    citation_id: str
    in_text_mark: str
    quote_body: str
    bibliography_line: str
    fetch_status: str = "ok"
    source_label: str = ""
    title: str = ""


class LiteratureInsertQuotesOut(BaseModel):
    quotes: list[LiteratureQuoteBlockOut]
    citations: list[CitationOut] = []


class LiteratureFormatIn(BaseModel):
    paper: dict
    style: str = "apa"
    index: int | None = None

    @field_validator("style")
    @classmethod
    def style_ok(cls, v: str) -> str:
        allowed = {"apa", "mla", "chicago", "gb_t7714"}
        if v not in allowed:
            raise ValueError("style must be one of apa, mla, chicago, gb_t7714")
        return v


class LiteratureFormatOut(BaseModel):
    citation: str

File: frontend/src/types/literature.ts
export interface LiteraturePaper {
  title: string;
  year?: number | null;
  authors: string[];
  journal?: string;
  doi?: string;
  citations?: number;
  type?: string | null;
  source?: string | null;
  source_label?: string | null;
  url?: string;
  semantic_scholar_id?: string | null;
  external_id?: string | null;
  abstract_preview?: string | null;
  document_type?: string | null;
  publisher?: string | null;
  volume?: string | null;
  issue?: string | null;
  pages?: string | null;
  quotable_snippet?: string | null;
}

export interface LiteratureSearchResult {
  papers: LiteraturePaper[];
  github: LiteraturePaper[];
  wiki: LiteraturePaper[];
  official_docs: LiteraturePaper[];
  refined_queries: string[];
  warnings?: string[];
  items: LiteraturePaper[];
  profile?: string;
  source_hint?: string;
}

export interface LiteratureRefineResult {
  refined_queries: string[];
  must_include: string[];
  must_exclude: string[];
}

export type LiteratureTab = "papers" | "github" | "wiki" | "official_docs";

export interface LiteratureQuoteBlock {
  citation_id: string;
  in_text_mark: string;
  quote_body: string;
  bibliography_line: string;
  fetch_status: string;
  source_label?: string;
  title?: string;
}

export function literaturePaperKey(p: LiteraturePaper): string {
  if (p.external_id && p.source) return `${p.source}:${p.external_id}`.toLowerCase();
  if (p.doi?.trim()) return `doi:${p.doi.toLowerCase()}`;
  return (p.title || "").toLowerCase();
}

export function literaturePaperUrl(p: LiteraturePaper): string {
  if (p.url?.trim()) return p.url.trim();
  if (p.source === "wikipedia" && p.external_id) {
    return `https://zh.wikipedia.org/wiki/${encodeURIComponent(p.external_id.replace(/ /g, "_"))}`;
  }
  if (p.source === "arxiv" && p.external_id) {
    return `https://arxiv.org/abs/${p.external_id}`;
  }
  if (p.source === "github" && p.external_id) {
    return `https://github.com/${p.external_id}`;
  }
  if (p.source === "official_doc" && p.url?.trim()) {
    return p.url.trim();
  }
  if (p.doi?.trim()) return `https://doi.org/${p.doi.replace(/^https?:\/\/doi\.org\//i, "")}`;
  if (p.semantic_scholar_id?.trim()) {
    return `https://www.semanticscholar.org/paper/${p.semantic_scholar_id}`;
  }
  if (p.title?.trim()) {
    return `https://scholar.google.com/scholar?q=${encodeURIComponent(p.title)}`;
  }
  return "";
}

export interface CitationRecord {
  id: string;
  book_id: string;
  doi?: string | null;
  title: string;
  authors: string[];
  year?: number | null;
  journal?: string | null;
  format_cache?: Record<string, string> | null;
  source: "literature_search" | "uploaded_file" | "manual";
  source_file_id?: string | null;
  raw_text?: string | null;
  quotable_snippet?: string | null;
  abstract_preview?: string | null;
  url?: string | null;
  document_type?: string | null;
  publisher?: string | null;
  volume?: string | null;
  issue?: string | null;
  pages?: string | null;
  metadata_status?: "complete" | "needs_completion";
  external_source?: string | null;
  list_index?: number | null;
  formatted?: string | null;
  created_at: string;
}
